core/templates/components/ck-editor-helpers/ck-editor-copy-content-service.ts

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for copying ck editor content from an output view to an
+ *               ck editor
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable, EventEmitter } from '@angular/core';
+
+import { HtmlEscaperService } from 'services/html-escaper.service';
+import { Subscription } from 'rxjs';
+
+interface CkEditorCopyEvent {
+  rootElement: HTMLElement;
+  containedWidgetTagName?: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CkEditorCopyContentService {
+  private copyEventEmitter = new EventEmitter<CkEditorCopyEvent>();
+  private ckeditorToSubscription: {[id: string]: Subscription} = {};
+
+  copyModeActive = false;
+
+  constructor(private htmlEscaperService: HtmlEscaperService) {}
+
+  /**
+   * Traverses up and down element ancestors/descendants, searching for widget
+   * tags.
+   * @param {HTMLElement} target The element target that contains or is a
+   *  descendant of a widget, or a plain HTML element/group.
+   * @returns {CkEditorCopyEvent} Returns an object of type CkEditorCopyEvent,
+   *  where rootElement is the root ancestor of target and is
+   *  always a child of angular-html-bind. containedWidgetTagName is the tag
+   *  name of the widget if found in ancestors or descendants.
+   */
+  private _handleCopy(target: HTMLElement): CkEditorCopyEvent {
+    let containedWidgetTagName;
+    let currentElement = target;
+
+    while (
+      currentElement.parentElement.tagName !== 'ANGULAR-HTML-BIND'
+    ) {
+      const currentTagName = currentElement.tagName.toLowerCase();
+      if (currentTagName.includes('-noninteractive-')) {
```

Comment:
@DubeySandeep this may be difficult to do, because widgets can contain any amount of tags. For example, a math widget contains spans, mathjax-bind, and no-br, while an image widget contains figures, divs, imgs, and figcaptions. 

Also, if we were to create a list of tags to not be copied, should it apply to what they click? Or the root element of what they click? Or the descendants?

For now, I think I will add a whitelist for (1) html tag root elements and (2) widgets. The first whitelist will only apply if it is not a widget. Does that sound good?

Reply:
Sorry for the misunderstanding, I think I was thinking of having a list of widget's tagnames like `oppia-noninteractive-images` etc and only allow copying these elements (i.e, their outerHtml), Will that be good, wdyt?

Reply:
Sure - added.